repentogon/LuaInterfaces/HUD/LuaMinimap.cpp

```
@@ -12,6 +14,257 @@
+			ANM2* curseSprite = entry.second.customANM2;
+
+			if (adjustedCurseId >= (MAX_CURSES + 1) && curseSprite->_loaded && (curses & (1U << (adjustedCurseId - 1)))) {
+				TryAddIcon(icons_to_render, num_icons, curseSprite, curses_anim, entry.second.frameNum);
+
+				if (num_icons >= MAX_MINIMAP_ICONS) break;
+			}
+		}
+	}
+
+	TryAddAllIcons(icons_to_render, num_icons, active_custom_icons_it, ICON_POSITION_AFTER_CURSE_ICONS);
+#pragma endregion
+
+#pragma region Rendering
+	Vector offset(-BASE_ICON_X_OFFSET, 0);
+	if (num_icons >= 4) {
+		offset += Vector(num_icons - 3.0f, 0);
+	}
+
+	for (size_t i = 0; i < num_icons; i++) {
+		MinimapIcon icon = icons_to_render[i];
+
+		icon.sprite->SetFrame(&icon.anim, icon.frame);
+
+		icon.sprite->_color._tint[3] = alpha;
+		icon.sprite->Render(&render_position, &vector_zero, &vector_zero);
+
+		render_position += offset;
+	}
+#pragma endregion
+}
+
+LUA_FUNCTION(Lua_MinimapRegisterCustomIcon) {
+	int position = (int)luaL_checkinteger(L, 1);
+	const char* anm2 = luaL_checkstring(L, 2);
+	std_string anm2_string(anm2);
+	const char* anim = luaL_checkstring(L, 3);
+	std_string anim_string(anim);
+	int frame = (int)luaL_checkinteger(L, 4);
+
+	RegisteredMinimapIcon* minimapIcon = new RegisteredMinimapIcon;
+	minimapIcon->position = position;
+
+	minimapIcon->sprite.Load(anm2_string, true);
+
+	if (!ValidateSprite(minimapIcon->sprite, anim, frame)) {
+		lua_pushnil(L);
+	} else {
+		minimapIcon->sprite.SetFrame(&anim_string, frame);
+
+		minimapIcon->id = RegisteredCustomIcons.size();
+		int index_to_place_at = 0;
+
+		auto lb = std::lower_bound(RegisteredCustomIcons.begin(), RegisteredCustomIcons.end(), minimapIcon, CompareIcons);
+		RegisteredCustomIcons.insert(lb, minimapIcon);
+
+		lua_pushinteger(L, minimapIcon->id);
+	}
+
+	return 1;
+}
+
+LUA_FUNCTION(Lua_MinimapActivateCustomIcon) {
+	int id = (int)luaL_checkinteger(L, 1);
+	
+	if (id < 0 || id >= RegisteredCustomIcons.size()) {
+		return 0;
```

This and other Lua functions that need input validation other than type, should trigger a lua arg error, and not silently fail.